### DISCUSSION TITLE
[Proposal]: A conditionless switch operator with a whole default statement

### DISCUSSION THREAD

**Discussion by c0debeef** [dotnet/csharplang#7358] [Language Ideas] [answered]
It would be nice to add the ability to use the switch statement without a condition, where the whole body is a default code fragment.

Sometimes I have to write complex checks that, if it fails, should completely break the execution of the code block. The use of exceptions is resource-intensive, and Dijkstra's precepts do not allow the goto operator.

Now the variant with a switch looks ugly:

```C#
switch (0)
{
    default:
    {
        // some stuff
        break;

        // some stuff
        break;

        // some stuff
        break;
    }
}
```

So, I have to cheat and write something like:

```C#
do
{
    // some stuff

    if (0 == Environment.TickCount % 2) break;

    // some stuff

    if (Regex.Match(Environment.TickCount.ToString(), "^[0|1]+$").Success) break;

    // success

} while (false); // switch
```

It would be nice if more native syntax could be used for this:

```C#
switch
{
    // some stuff

    if (0 == Environment.TickCount % 2) break;

    // some stuff

    if (Regex.Match(Environment.TickCount.ToString(), "^[0|1]+$").Success) break;

    // success
}
```

**Comment by erik-kallen** [comment]
If you want to use a goto, just write goto. Hiding it behind other constructs that mean the same thing is just useless.

**Reply by c0debeef** [reply]
I don't want to use goto.

**Reply by TahirAhmadov** [reply]
Why don't you? C# goto is not your grandfather's goto.

**Comment by CyrusNajmabadi** [comment]
I have no idea why a 'goto' would not perfectly appropriate here.

**Reply by c0debeef** [reply]
A goto operator is considered harmful and not well suited for high level programming language.
It's like an asm jump to label instruction which difficult for human to maintain and understand.

**Reply by CyrusNajmabadi** [reply]
`goto` is not harmful (which is why we have it in the language, and support its usage).  We're opposed to proposals that serve to just replicate it, just with slightly different syntax.

**Reply by c0debeef** [reply]
This is not a reproduction of goto statement since it represents a jump to a label anywhere in the program.
This is a simplification of the switch operator syntax.

**Reply by 333fred** [reply]
> it represents a jump to a label anywhere in the program

C# is not C. `goto` cannot jump to any label anywhere in the current file.

**Reply by c0debeef** [reply]
Goto is poorly suited for this case in that in order to understand the code with it, you need to know exactly where the transition is made, whether there are other transition markers, and which code block should serve this goto if you have to modify the code.

When using conditionless switch or loop, we automatically know both the code block and where the exit occurs when the break operator is reached, and also that this is the only certain way to exit the block.
This approach is comparable to the reasons for using for(;;) {} or while(true) {}.

**Reply by TahirAhmadov** [reply]
You can create a scope with curly braces, add a label at the end of it or right after it, then goto that label.

**Comment by c0debeef** [comment]
The goto operator should normally be used in exceptional cases. For example, to exit from many nested loops, if this cannot be done efficiently by other methods.

**Reply by CyrusNajmabadi** [reply]
> The goto operator should normally be used in exceptional cases. 

That is not hte philosophy of our language.  A goto should be used when you want to unconditionally jump to a specific location (like in your OP) :)

**Reply by DavidArno** [reply]
If you do not wish to use `goto` (and that is a position I can understand), then use `return`. Make that block a method body and return from the method when a check fails.

**Comment by c0debeef** [comment]
Then you folks should share your resumes that mention C# programming using goto statement :)

**Reply by CyrusNajmabadi** [reply]
Fred and I are two of the language designers for C# :)

**Reply by koszeggy** [reply]
Maybe you should have a look at the runtime sources. `goto` is heavily used in performance critical places eg. [here](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Marvin.cs,797a1d788badc2e7,references) or [here](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Buffer.cs,a02c7200e8b6cc36,references).

I assume your gotophobia is coming from another language where using it could be dangerous and wasn't just a readability issue. In C# you cannot jump in or out of a method, `try` or `using` block, or to a line where a variable is not initialized.

And in fact `goto` may even help readability in special cases. For example, C# doesn't let you just omit the `break` in a `switch` statement so you must explicitly express your intention to fall though `case` labels by a `goto`, whereas in other languages omitting the `break` usually was just an oversight, which led to quite annoying bugs.

**Reply by c0debeef** [reply]
I was writing sorts of examples with goto and I had a question.
In this example, is it a bug or a feature, and how obvious is the behavior of goto in this case?

```C#
static void Main(string[] args)
{
    string foo = "bar";

    switch (foo)
    {
        case null:
            Debug.Write("null");
            goto case default;
        default:
            Debug.Write("default");
            goto case null;

    }
}
```

```C#
static void Main(string[] args)
{
    string foo = "bar";

    switch (foo)
    {
        case null:
            Debug.Write("null");
            goto default;
        default:
            Debug.Write("default");
            goto case null;

    }
}
```

**Reply by koszeggy** [reply]
Seems as if you were waiting for such a response. Btw. since both examples are infinite loops (just a bit differently) I cannot really decide if one of them meant to be more 'correct' than the another.

Anyhow, the explicit `goto`s make it clear and easily readable that for some reason you want to jump even from `default` to another case, which should be immediately suspicious just by looking at the code, even without spotting the `goto case default` part (I guess that meant to be the point of your example).

But intentionally bad examples can be created for many language constructs and they do not lead anywhere. Here is my favorite one:

```cs
public static void Main() => SerializeCollection(new sbyte[] { 1, -2, 3, -4 });

private static void SerializeCollection<T>(IEnumerable<T> coll)
{
    // shortcut for byte array, which has a special support/format in our serializer
    if (coll is byte[] bytes)
    {
        Console.WriteLine($"Processing bytes: {String.Join(", ", bytes)}");
        return;
    }

    // other collections are serialized by indicating type name and using another format
    Console.WriteLine($"Processing other type of collection: {String.Join(", ", coll)}");
}
```

Or is it a bug or feature that `coll is IEnumerable<T> c` can be refactored to `coll is {} c` but not to `coll is var c` because the latter is incompatible with the first one (captures also `null`)?

Do the examples above mean that we should ban the `is` operator?

I repeat, `goto` **may** even help readability, even though it can also be used for other reasons, eg. performance. Every tool has its purpose and can be used badly or well.

See also (not C#-specific): https://softwareengineering.stackexchange.com/questions/566/is-using-goto-ever-worthwhile

**Reply by c0debeef** [reply]
In the original post, I wrote that I was using a while bypass, but I mentioned goto in passing. After that, everyone started telling me how beautiful goto is.

Btw, in the first example, `goto case default` surprisingly means `goto case default(string) ` i.e. `null `

**Reply by koszeggy** [reply]
No one says how beautiful `goto` is. It's just a reasonable choice for your problem. And it's you who want to fire developers who mention the usage of `goto` in their resume.

> Btw, in the first example, `goto case default` surprisingly means `goto case default(string)` i.e. `null` 

That's what I was referring to in my answer (_"even without spotting the `goto case default` part"_). And I also mentioned similar non-`goto` constructs with surprising outcomes, suggesting that many tools can be misused and misusing a tool doesn't necessarily mean that it's inherently bad.

**Comment by c0debeef** [comment]
This is a proposal. Why then is there a switch statement if we can write everything with goto?

**Reply by CyrusNajmabadi** [reply]
The ability to have the compiler match a particular value against a constant or pattern and jump to that automatically.

**Reply by c0debeef** [reply]
Maybe instead of switch you should use a new keyword like block. But why can methods be parameters, but a switch cannot?

**Comment by c0debeef** [comment]
Perhaps the main reason people don't want to use goto is if someone sees your code and informs the relevant authorities where they'll come after you and shoot immediately.

**Reply by iam3yal** [reply]
Who cares what other people think? the language designers told you that your assumptions are incorrect and you can use `goto` for the case you provided in the OP that's the only important things that you should take and do yourself a favour and speak for yourself! what other people do or not is not important to get your problems solved, unless they are part of the problem or the solution. 😄

**Reply by c0debeef** [reply]
My gotophobia is probably from the experience of programming and reading all sorts of smart books on it. For example, in the book that Anders Hejlsberg wrote, there are such [annotations](https://ibb.co/0ZH6553).

**Reply by HaloFour** [reply]
Advice is different from doctrine.  Even in Dijkstra's paper "Go To Considered Harmful" he outlines what he believes to be perfectly suitable use cases for it.  One should strive to write readable and maintainable code, and it's amusing to me how developers can sometimes make their code much less legible when they bend over backwards to avoid language features like `goto`.

**Reply by HaloFour** [reply]
Dijkstra also thought that repeating/looping clauses were superfluous since recursion was simpler, yet that bit of advice doesn't seem to have caught on with the same veracity.

**Reply by iam3yal** [reply]
@c0debeef I don't really care about _smart books_ or _who wrote them_ if there is something I can use that solves the problem today and makes my code easy to read and follow I'd use it! and wouldn't punish myself and others with workarounds that are anything but readable only because it was written somewhere, there is always an exception to the rule so when you follow the rule blindly and when the workarounds are making the code hurt your eyes it's probably a good time to make a judgment and break out of the rule you follow so enthusiastically. there's nothing wrong using `goto` sparingly where it makes sense.

**Comment by mattwar** [comment]
There is a new statement, 'goneto'.   You use it to leave a note to the next thread running through the code to tell it where the last thread went.

**Reply by HaloFour** [reply]
And don't forget `comefrom`.

**Reply by c0debeef** [reply]
And as many people definitely know - goto in the code spreads its magic to other operators around)
[[V3020](https://pvs-studio.com/en/docs/warnings/v3020/) An unconditional 'return' within a loop. Enumerable.cs 517](https://github.com/dotnet/runtime/issues/15993#issuecomment-167845742)

**Comment by Richiban** [comment, marked as answer]
Just use a local function; then you have the `return` keyword. 

```cs
void Stuff()
{
    // some stuff

    if (0 == Environment.TickCount % 2) return;

    // some stuff

    if (Regex.Match(Environment.TickCount.ToString(), "^[0|1]+$").Success) return;

    // success

} 

Stuff();

// Code outside the local function
```

**Reply by c0debeef** [reply]
In practice, we can write code in any way, but syntactic sugar makes life easier and we can ask for it.

**Reply by c0debeef** [reply]
If the break in the switch operator were not a vestige, then we could write something like this without lovely goto:
```C#
switch (state)
{
    case 0:
        _enumerator = _source.GetEnumerator();
        continue;
    case 1:
        _enumerator.MoveNext();
        continue;
    case 2:
        //etc..;
        break;
}
```

**Reply by HaloFour** [reply]
You already have many control structures that you can use to accomplish the same behavior, if you're intentionally avoiding `goto`.  You reference them in your original proposal.  Adding new syntactic forms for every possible permutation and use case is not a sustainable strategy.

**Reply by c0debeef** [reply]
Language develops not only by making colossal changes, right? I don't think the parameterless switch makes breaking changes. Take for example the same `default(T) => default`.

**Reply by CyrusNajmabadi** [reply]
We make changes if we think the value is there. Inventing a clone of 'goto' is not something we think is valuable. As mentioned, just have a block, a label, and gotos to go to that label.  The workaround is simple and easy to understand.
